Kubernetes end to end tests should run on Macs.

Currently the tests assume that you are using Linux. This allows them to build Vector locally and deploy a container to test to Docker Hub. 

On a Mac you have to deploy the Vector container to Docker Hub yourself and then specify `SKIP_CONTAINER_IMAGE_PUBLISHING=true CONTAINER_IMAGE=<image>` when running the tests. We should build it into the tests to use Docker to build a Vector image for Macs. There is also a possibility we could use cross compiling, but I couldn't get it to work - there were issues with some of the third party libraries we use.

I received this error after several hourse of building: 

```
error: could not compile `k8s-openapi`

Caused by:
  process didn't exit successfully: `rustc --crate-name k8s_openapi --edition=2018 /cargo/registry/src/github.com-1ecc6299db9ec823/k8s-openapi-0.11.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 -C linker-plugin-lto -C codegen-units=1 --cfg 'feature="api"' --cfg 'feature="default"' --cfg 'feature="http"' --cfg 'feature="percent-encoding"' --cfg 'feature="url"' --cfg 'feature="v1_16"' -C metadata=5611f29a59d6d2dc -C extra-filename=-5611f29a59d6d2dc --out-dir /target/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/target/x86_64-unknown-linux-gnu/release/deps -L dependency=/target/release/deps --extern base64=/target/x86_64-unknown-linux-gnu/release/deps/libbase64-6061d76424c3816a.rmeta --extern bytes=/target/x86_64-unknown-linux-gnu/release/deps/libbytes-cd5092bb6fd0e2f2.rmeta --extern chrono=/target/x86_64-unknown-linux-gnu/release/deps/libchrono-201f9b240215be4d.rmeta --extern http=/target/x86_64-unknown-linux-gnu/release/deps/libhttp-fb1421e8e1952222.rmeta --extern percent_encoding=/target/x86_64-unknown-linux-gnu/release/deps/libpercent_encoding-45dc6f248b908bcb.rmeta --extern serde=/target/x86_64-unknown-linux-gnu/release/deps/libserde-76586474eebda37b.rmeta --extern serde_value=/target/x86_64-unknown-linux-gnu/release/deps/libserde_value-9b0283a13b017c04.rmeta --extern serde_json=/target/x86_64-unknown-linux-gnu/release/deps/libserde_json-7d5d8d6dc64a6c50.rmeta --extern url=/target/x86_64-unknown-linux-gnu/release/deps/liburl-0101410dca9c8a4d.rmeta --cap-lints allow -C link-args=-rdynamic` (signal: 9, SIGKILL: kill)
warning: build failed, waiting for other jobs to finish...
error: build failed
make[1]: *** [Makefile:238: target/x86_64-unknown-linux-gnu/release/vector] Error 101
make[1]: Leaving directory '/Users/stephen.wakely/src/vector'
make: *** [Makefile:495: test-e2e-kubernetes] Error 2
```

Am trying to build again, but it seems exceedingly slow (at least several hours). Should that be the case?

--- a/.github/workflows/k8s_e2e.yml
+++ b/.github/workflows/k8s_e2e.yml
@@ -95,7 +95,7 @@ jobs:
           script: |
             // Parameters.
             const minikube_version = [
-              "v1.15.1",
+              "v1.21.0",
             ]
             const kubernetes_version = [
               { version: "v1.19.2", is_essential: true },

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -826,15 +826,16 @@ Vector release artifacts are prepared for E2E tests, so the ability to do that
 is required too, see Vector [docs](https://vector.dev) for more details.
 
 Notes:
+
 > - `minikube` had a bug in the versions `1.12.x` that affected our test
-> process - see https://github.com/kubernetes/minikube/issues/8799.
-> Use version `1.13.0+` that has this bug fixed.
+>   process - see https://github.com/kubernetes/minikube/issues/8799.
+>   Use version `1.13.0+` that has this bug fixed.
 > - `minikube` has troubles running on ZFS systems. If you're using ZFS, we
-> suggest using a cloud cluster or [`minik8s`](https://microk8s.io/) with local
-> registry.
+>   suggest using a cloud cluster or [`minik8s`](https://microk8s.io/) with local
+>   registry.
 > - E2E tests expect to have enough resources to perform a full Vector build,
-> usually 8GB of RAM with 2CPUs are sufficient to succesfully complete E2E tests
-> locally.
+>   usually 8GB of RAM with 2CPUs are sufficient to succesfully complete E2E tests
+>   locally.
 
 ##### Running the E2E tests
 

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -825,15 +825,16 @@ E2E (end-to-end) tests.
 Vector release artifacts are prepared for E2E tests, so the ability to do that
 is required too, see Vector [docs](https://vector.dev) for more details.
 
-> Note: `minikube` had a bug in the versions `1.12.x` that affected our test
+Notes:
+> - `minikube` had a bug in the versions `1.12.x` that affected our test
 > process - see https://github.com/kubernetes/minikube/issues/8799.
 > Use version `1.13.0+` that has this bug fixed.
-
-Also:
-
-> Note: `minikube` has troubles running on ZFS systems. If you're using ZFS, we
+> - `minikube` has troubles running on ZFS systems. If you're using ZFS, we
 > suggest using a cloud cluster or [`minik8s`](https://microk8s.io/) with local
 > registry.
+> - E2E tests expect to have enough resources to perform a full Vector build,
+> usually 8GB of RAM with 2CPUs are sufficient to succesfully complete E2E tests
+> locally.
 
 ##### Running the E2E tests
 

--- a/scripts/test-e2e-kubernetes.sh
+++ b/scripts/test-e2e-kubernetes.sh
@@ -20,12 +20,18 @@ random-string() {
 
 # Detect if current kubectl context match the one from `minikube`.
 is_kubectl_context_minikube() {
-  [[ "$(kubectl config current-context || true)" == "$(minikube profile)" ]]
+  MINIKUBE_CLUSTER=$(minikube profile || true)
+  [[ "$(kubectl config current-context || true)" == "${MINIKUBE_CLUSTER:-"minikube"}" ]]
 }
 
 # Detect if current kubectl context is `kind`.
 is_kubectl_context_kind() {
-  [[ "$(kubectl config current-context || true)" == "kind-kind" ]]
+  for KIND_CLUSTER in $(kind get clusters || true); do
+    if [[ "$(kubectl config current-context || true)" == "kind-${KIND_CLUSTER}" ]]; then
+      return 0
+    fi
+  done
+  return 1
 }
 
 

--- a/scripts/test-e2e-kubernetes.sh
+++ b/scripts/test-e2e-kubernetes.sh
@@ -18,9 +18,9 @@ random-string() {
   echo
 }
 
-# Detect if current kubectl context is `minikube`.
+# Detect if current kubectl context match the one from `minikube`.
 is_kubectl_context_minikube() {
-  [[ "$(kubectl config current-context || true)" == "minikube" ]]
+  [[ "$(kubectl config current-context || true)" == "$(minikube profile)" ]]
 }
 
 # Detect if current kubectl context is `kind`.
@@ -121,8 +121,8 @@ fi
 if [[ -z "${SKIP_CONTAINER_IMAGE_PUBLISHING:-}" ]]; then
   # Make the container image accessible to the k8s cluster.
   if is_minikube_cache_enabled; then
-    minikube cache add "$CONTAINER_IMAGE"
-    trap 'minikube cache delete "$CONTAINER_IMAGE"' EXIT
+    minikube image load "$CONTAINER_IMAGE"
+    trap 'minikube image rm "$CONTAINER_IMAGE"' EXIT
   else
     docker push "$CONTAINER_IMAGE"
   fi

--- a/scripts/test-e2e-kubernetes.sh
+++ b/scripts/test-e2e-kubernetes.sh
@@ -18,13 +18,13 @@ random-string() {
   echo
 }
 
-# Detect if current kubectl context match the one from `minikube`.
+# Detect if current kubectl context matches the one from `minikube`.
 is_kubectl_context_minikube() {
   MINIKUBE_CLUSTER=$(minikube profile || true)
   [[ "$(kubectl config current-context || true)" == "${MINIKUBE_CLUSTER:-"minikube"}" ]]
 }
 
-# Detect if current kubectl context is `kind`.
+# Detect if current kubectl context matches a `kind` cluster.
 is_kubectl_context_kind() {
   for KIND_CLUSTER in $(kind get clusters || true); do
     if [[ "$(kubectl config current-context || true)" == "kind-${KIND_CLUSTER}" ]]; then

